File: RainWaterTrap.java
/**
 * Time Complexity : O(n) where n is the number of elements 
 * Space Complexity : O(1)
 */
public class RainWaterTrap {
    public int trap(int[] height) {
        if(height == null || height.length == 0){return 0;}
        int lw  =0, rw =0, l =0, r = height.length-1, amount  = 0;
        while(l <= r){
            if(lw < rw){                                                          
                if(height[l] < lw){                                             
                    amount += lw - height[l];                                 
                }
                else {
                    lw = height[l];                                                     
                }
                l++;
            } else {
                 if(height[r] < rw){
                    amount += rw - height[r];                                                   
                }
                else {
                    rw = height[r];                                                             
                }
                r--;
            }
        }
        return amount;                                                                              
    } 
}

File: Rain Water Trap.py
class Solution:
    def trap(self, height: List[int]) -> int:
        leftHighest = [0]*len(height)
        rightHighest = [0]*len(height)
        
        curr = 0
        sumArray = [0]*len(height)
        for i in range(len(leftHighest)):
            leftHighest[i] = curr
            curr = max(curr, height[i])
        curr = 0
        result = 0
        for i in range(len(rightHighest)-1, -1, -1):
            rightHighest[i] = curr
            sumArray[i] = max(min(leftHighest[i], rightHighest[i])-height[i],0)
            result+=sumArray[i]
            curr = max(curr, height[i])

        return result
Time: O(N)
Space: O(N)
            
